airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/S3ConsumerFactory.java

```diff
@@ -0,0 +1,169 @@
+        onCloseFunction(storageOperations, writeConfigs),
+        catalog,
+        storageOperations::isValidData);
+  }
+
+  private static List<WriteConfig> createWriteConfigs(final BlobStorageOperations storageOperations,
+                                                      final NamingConventionTransformer namingResolver,
+                                                      final JsonNode config,
+                                                      final ConfiguredAirbyteCatalog catalog) {
+    return catalog.getStreams()
+        .stream()
+        .map(toWriteConfig(storageOperations, namingResolver, config))
+        .collect(Collectors.toList());
+  }
+
+  private static Function<ConfiguredAirbyteStream, WriteConfig> toWriteConfig(
+                                                                              final BlobStorageOperations storageOperations,
+                                                                              final NamingConventionTransformer namingResolver,
+                                                                              final JsonNode config) {
+    return stream -> {
+      Preconditions.checkNotNull(stream.getDestinationSyncMode(), "Undefined destination sync mode");
+      final AirbyteStream abStream = stream.getStream();
+      final String namespace = abStream.getNamespace();
+      final String streamName = abStream.getName();
+      final String outputNamespace = getOutputNamespace(abStream, config.get("s3_bucket_path").asText(), namingResolver);
+      final String customOutputFormat =
+          config.has("s3_path_format") && !config.get("s3_path_format").asText().isBlank() ? config.get("s3_path_format").asText()
+              : S3DestinationConstants.DEFAULT_PATH_FORMAT;
+      final String outputBucketPath = storageOperations.getBucketObjectPath(outputNamespace, streamName, SYNC_DATETIME, customOutputFormat);
+      final DestinationSyncMode syncMode = stream.getDestinationSyncMode();
+      final WriteConfig writeConfig = new WriteConfig(namespace, streamName, outputNamespace, outputBucketPath, syncMode);
+      LOGGER.info("Write config: {}", writeConfig);
+      return writeConfig;
+    };
+  }
+
+  private static String getOutputNamespace(final AirbyteStream stream,
+                                           final String defaultDestNamespace,
+                                           final NamingConventionTransformer namingResolver) {
+    return stream.getNamespace() != null
+        ? namingResolver.getNamespace(stream.getNamespace())
+        : namingResolver.getNamespace(defaultDestNamespace);
+  }
+
+  private OnStartFunction onStartFunction(final BlobStorageOperations storageOperations, final List<WriteConfig> writeConfigs) {
+    return () -> {
+      LOGGER.info("Preparing bucket in destination started for {} streams", writeConfigs.size());
+      for (final WriteConfig writeConfig : writeConfigs) {
+        final String namespace = writeConfig.getNamespace();
```

Comment:
nit: these declarations should probably be inside the following if block, because they are not used outside of it